### PR TITLE
feat(core/biblio): support different bibliographical styles

### LINKED ISSUE #857: Support different bibliographical styles
The publisher is listed after the Authors, which is not a common bibliographic style.
Can the publisher item be moved to the end (before the URL ref)?

### PR DESCRIPTION
**Pull request by OrionStar25**
Fixes #857. 

Adds support for APA, MLA, and retains ReSpec's custom citation style as the default.

### CODE REVIEW

**Comment by OrionStar25** [comment]
Should I move on to writing a test @marcoscaceres?

**Comment by marcoscaceres** [comment]
Yes please! Let me know what you plan to tests. Happy to provide your with additional ideas.

**Comment by OrionStar25** [comment]
I have a doubt. Right now what happens is that wire Reference creates an Html frame for our stringifyReference strings to fit into.
But now we are giving an option to make that HTML frame flexible, so shouldn't wireReference be changed too?

**Comment by OrionStar25** [comment]
Also, localBiblio should now have a "style" configuration present.  
Checking biblio-spec for what can be done.

**Comment by OrionStar25** [comment]
I was thinking  "style" should be sent to wireReference and then according to the chosen style, an appropriate HTML frame is returned. Then stringifyReference adds our output string in the frame.
Does that make sense?

**Comment by marcoscaceres** [comment]
Ah, so instead of `localBiblio` having the configuration option, we should just have a global config option `biblioStyle`.

**Comment by marcoscaceres** [comment]
global meaning `conf.biblioStyle`

**Comment by OrionStar25** [comment]
Oh yes that's better. Alright so 2 extra test cases shall be added in biblio-spec.js for APA and MLA. Then I'll check if the chosen style is only getting implemented. I'm.not quite sure if I should be rewriting any assertion mentioned there, could you please help me with that?

**Comment by marcoscaceres** [comment]
Only write new tests cases, as all existing test cases should continue to work exactly as before. If any existing ones fail, then that’s a bug 🐛

**Comment by OrionStar25** [comment]
@marcoscaceres , I used `push --force` hence had to reopen this. I didn't add a test for MLA as the structure for MLA is being tested twice already (when publisher is present). I couldn't find where we define our `global conf`, but I did pass it as a parameter.

File: tests/spec/core/biblio-spec.js [hunk 1]
```diff
@@ -102,6 +107,18 @@ describe("W3C — Bibliographic References", () => {
     expect(ref.textContent).toMatch(/Publisher Here\.\s/);
   });
 
+  it("displays the reference for APA citation", async() => {
+    // Make sure the reference is added.
+    let ref = doc.querySelector("#bib-TestRef4 + dd");
+    expect(ref).toBeTruthy();
+    // This prevents Jasmine from taking down the whole test suite if SpecRef is down.
+    if (!specRefOk) {
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: We shouldn't need this here, as this is a local reference.

File: src/core/biblio.js [hunk 2]
```diff
@@ -93,24 +93,76 @@ export function wireReference(rawRef, target = "_blank") {
   `;
 }
 
-export function stringifyReference(ref) {
+// Author, A. (Year, Month Date of Publication). Article title. Retrieved from URL
+function entryToAPA(ref) {
```
Comment (marcoscaceres) [inline src/core/biblio.js]: So, based on @saschanaz's feedback, we should just be treating these as HTML objects, and not text... so we need refactor these like so:

```JS
const html = (...args) => hyperHTML.bind(doc.createDocumentFragment())(...args)

function entryToAPA(ref) {
  const { title } = ref;
  const authors =
    ref.authors && ref.authors.length
      ? `${ref.authors.join("; ")} ${ref.etAl ? " et al" : ""}.`
      : "";
  const date = ref.date ? `(${ref.date})` : "";
  return html`
    ${authors} ${date} <a href="${ref.href}"><cite>${title}</cite></a>
    ${ref.href ? html`Retrieved from URL: ${html`<a href="${ref.href}">${ref.href}</a>`}` : ""}
  `;
}

function entryToMLA(ref) {
   // as above, but for MLA
}

function entryToW3C(ref) {
  // as above, but for W3C...
}
```

The we can just add them directly to the document as real HTML elements.

File: tests/spec/core/biblio-spec.js (hunk 1 above)
Comment (OrionStar25) [reply tests/spec/core/biblio-spec.js]: Alright, so no assertions needed?

Comment (marcoscaceres) [reply tests/spec/core/biblio-spec.js]: just the `if (!specRefOk) { ... ` can go away, as it doesn't serve any useful purpose here. `specRefOk` is for checking if we can talk to the Specref server - but here, we don't talk to the server.

**Comment by OrionStar25** [comment]
@marcoscaceres I'm stuck :disappointed:  . I can't seem to get past these build errors. Also, its making biblio-db.js to fail.  I haven't changed `stringifyReference` yet, because I wanted the tests to pass first.

**Comment by OrionStar25** [comment]
@marcoscaceres

File: src/core/biblio.js [hunk 3]
```diff
@@ -93,24 +93,152 @@ export function wireReference(rawRef, target = "_blank") {
   `;
 }
 
-export function stringifyReference(ref) {
+// const html = (...args) => hyperHTML.bind(doc.createDocumentFragment())(...args);
```
Comment (marcoscaceres) [inline src/core/biblio.js]: We should probably get rid of this commented out code.

File: tests/spec/core/biblio-spec.js [hunk 4]
```diff
@@ -1,10 +1,11 @@
 "use strict";
 describe("W3C — Bibliographic References", () => {
-  const config = {
+  //Configuration for default of ReSpec
+  const conf1 = {
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: call this, "confDefault"

File: tests/spec/core/biblio-spec.js [hunk 5]
```diff
@@ -14,37 +15,147 @@ describe("W3C — Bibliographic References", () => {
...
       },
       BARBAR: {
-        title: "The BARBAR Spec",
+        title: "The BARBAR Spec"
+      }
+    }
+  };
+
+  //Configuration for APA citation
+  const conf2 = {
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: here, you can just do:

```JS
const confAPA = {
   ...confDefault,
   biblioStyle: "APA",
};

const confMLA = {
   ...confDefault,
   biblioStyle: "MLA",
};
```

So then you don't need all the duplicate code ✨

File: tests/spec/core/biblio-spec.js [hunk 6]
```diff
@@ -14,37 +15,147 @@ describe("W3C — Bibliographic References", () => {
...
+      FOOBARGLOP: {
+        aliasOf: "BARBAR"
+      },
+      BARBAR: {
+        title: "The BARBAR Spec"
+      }
+    }
+  };
+
+  const body1 = `
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: You should be able to use just 1 body, don't need 3. Each body can then have a different citation style, based on the config.

File: tests/spec/core/biblio-spec.js [hunk 7]
```diff
@@ -58,13 +169,43 @@ describe("W3C — Bibliographic References", () => {
...
+
   afterAll(flushIframes);
   const bibRefsURL = new URL("https://specref.herokuapp.com/bibrefs");
 
-  let doc;
+  let doc1, doc2, doc3;
   let specRefOk;
   beforeAll(async () => {
-    doc = await makeRSDoc({ config, body });
+    doc1 = await makeRSDoc({ conf1, body1 });
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: rename: "docDefault", "docMLA", "docAPA"... so it's clear which is which.

File: src/core/biblio.js (hunk 3 above)
Comment (marcoscaceres) [reply src/core/biblio.js]: Ah, sorry... this is the updated code... can you uncomment this, and remove the other stuff?

File: tests/spec/core/biblio-spec.js (hunk 6 above)
Comment (OrionStar25) [reply tests/spec/core/biblio-spec.js]: Alright, but how do I manage the change in number of tests cases? Or should I reduce all. To 3 test cases?

File: tests/spec/core/biblio-spec.js (hunk 5 above)
Comment (OrionStar25) [reply tests/spec/core/biblio-spec.js]: Spread the values, got it.

**Comment by OrionStar25** [comment]
@marcoscaceres made the requested changes.

File: tests/spec/core/biblio-spec.js [hunk 8]
```diff
@@ -36,6 +37,13 @@ describe("W3C — Bibliographic References", () => {
         href: "http://test.com",
         publisher: "Publisher Here",
       },
+      TestRef5: {
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: This should be remade to "RefWithOnlyHref"

File: tests/spec/core/biblio-spec.js [hunk 9]
```diff
@@ -36,6 +37,13 @@ describe("W3C — Bibliographic References", () => {
...
+        href: "http://test.com",
+      },
+      TestRef6: {
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: this one should be renamed to "RefWithOnlyHrefAndAuthor"

File: src/core/biblio.js [hunk 10]
```diff
@@ -154,6 +223,7 @@ function bibref(conf) {
         .appendTo($dl);
       var $dd = $("<dd></dd>").appendTo($dl);
       var refcontent = conf.biblio[ref];
+      var refstyle = conf.biblioStyle;
```
Comment (marcoscaceres) [inline src/core/biblio.js]: nit: 
```JS
const { biblioStyle: refstyle }  = conf;
```

File: tests/spec/core/biblio-spec.js [hunk 11]
```diff
@@ -92,26 +117,104 @@ describe("W3C — Bibliographic References", () => {
...
     expect(ref.textContent).toMatch(/Publisher Here\.\s/);
   });
 
+  it("displays the reference for APA citation", () => {
+    // Make sure the reference is added.
+    let ref = docAPA.querySelector("#bib-TestRef1 + dd");
+    expect(ref).toBeTruthy();
+    expect(ref.textContent).toMatch(/William Shakespeare\.\s/);
+    expect(ref.textContent).toMatch(/Test ref title\.\s/);
+    expect(ref.textContent).toMatch("Retrieved from URL: http://test.com/");
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: Ok, but now we want to test the finished product:
```JS
const finalString = "William Shakespeare. Test ref title. Retrieved from URL: http://test.com/";
expect(ref.textContent.trim()).toEqual(finalString); 
```

The same applies to each below.

Comment (marcoscaceres) [reply tests/spec/core/biblio-spec.js]: We also want to make sure that the right things are hyperlinked:

```JS
const anchors = ref.querySelectorAll("a");
expect(anchors.length).toEqual(2);
const [title, retrievedFrom] = [...anchors];
expect(title.href).toEqual("http://test.com/");
expect(title.firstElementChild.localName).toEqual("cite");
expect(title.firstElementChild.textContent).toEqual("Test ref title");
expect(retrievedFrom.href).toEqual("http://test.com/");
```

And so on... have a look at the code that created this, and make sure the HTML structure is what you expect it to be.

File: tests/spec/core/biblio-spec.js [hunk 12]
```diff
@@ -92,26 +117,104 @@ describe("W3C — Bibliographic References", () => {
...
+    ref = null;
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: If you find you need to reset things, then this signals you need a new test, so here:
```JS
it(".... describe what is different...", ()=>{

}); 
```

File: tests/spec/core/biblio-spec.js [hunk 13]
```diff
@@ -92,26 +117,104 @@ describe("W3C — Bibliographic References", () => {
...
+  it("displays the reference for MLA citation", () => {
+    // Make sure the reference is added.
+    let ref = docMLA.querySelector("#bib-TestRef1 + dd");
+    expect(ref).toBeTruthy();
+    expect(ref.textContent).toMatch(/William Shakespeare\.\s/);
+    expect(ref.textContent).toMatch(/"Test ref title\."\s/);
+    expect(ref.textContent).toMatch(/Publishers Inc,/);
+    expect(ref.textContent).toMatch("http://test.com");
+    ref = null;
+    // title,url
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: As above, but for MLA :)

**Comment by OrionStar25** [comment]
@marcoscaceres please review this!

**Comment by OrionStar25** [comment]
@marcoscaceres I hope that's what you meant!

File: src/core/biblio.js [hunk 14]
```diff
@@ -93,24 +93,93 @@ export function wireReference(rawRef, target = "_blank") {
...
-  }
-  if (ref.authors && ref.authors.length) {
-    output += ref.authors.join("; ");
-    if (ref.etAl) output += " et al";
-    output += ".";
+const html = (...args) => hyperHTML.bind(doc.createDocumentFragment())(...args);
+
+// Author, A. (Year, Month Date of Publication). Article title. Retrieved from URL
+function entryToAPA(ref) {
+  const { title } = ref.title;
```
Comment (marcoscaceres) [inline src/core/biblio.js]: Here it should be:

const { title } = ref;

Otherwise it wont work.

File: src/core/biblio.js [hunk 15]
```diff
@@ -93,24 +93,93 @@ export function wireReference(rawRef, target = "_blank") {
...
+    ref.authors && ref.authors.length
+      ? `${ref.authors.join("; ")} ${ref.etAl ? " et al" : ""}.`
+      : "";
+  const date = ref.date ? `(${ref.date})` : "";
+  return html`
+    ${authors} ${date} ${
+    ref.href
+      ? html`<a href="${ref.href}"><cite>${title}</cite></a>.`
+      : `${ref.title}`
   }
```
Comment (marcoscaceres) [inline src/core/biblio.js]: I think we should use a cite element here too.

File: src/core/biblio.js [hunk 16]
```diff
@@ -93,24 +93,93 @@ export function wireReference(rawRef, target = "_blank") {
...
+  const publisher = ref.publisher
+    ? `${ref.publisher} ${ref.publisher.endsWith(".")}`
+    : "";
+
+  return html`
+	${authors} ${
+    ref.href
+      ? html`<a href="${ref.href}"><cite>${title}</cite></a>.`
+      : `${title}`
+  } ${publisher} ${date} ${
```
Comment (marcoscaceres) [inline src/core/biblio.js]: As above regarding cite

File: src/core/biblio.js [hunk 17]
```diff
@@ -93,24 +93,93 @@ export function wireReference(rawRef, target = "_blank") {
...
+  const date = ref.date ? `${ref.date}. ` : "";
+  const status = ref.status
+    ? (REF_STATUSES.get(ref.status) || ref.status) + ". "
+    : "";
+
+  return html`
+  	${
+      ref.href
+        ? html`<a href="${ref.href}"><cite>${title}</cite></a>.`
+        : `${ref.title}`
```
Comment (marcoscaceres) [inline src/core/biblio.js]: As above

File: src/core/biblio.js [hunk 18]
```diff
@@ -165,13 +235,14 @@ function bibref(conf) {
         } else {
           key = refcontent.aliasOf;
           refcontent = conf.biblio[key];
+          refstyle = conf.biblioStyle;
```
Comment (marcoscaceres) [inline src/core/biblio.js]: I don’t think you need this line.

**Comment by OrionStar25** [comment]
@marcoscaceres Can you review the tests now please? :smiley:

File: tests/spec/core/biblio-spec.js [hunk 19]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
     expect(ref.textContent).toMatch(/Publishers Inc\.\s/);
     ref = null;
     // Make sure the ". " is automatically added to publisher.
-    ref = doc.querySelector("#bib-TestRef2 + dd");
+    ref = docDefault.querySelector("#bib-TestRef2 + dd");
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: Ok, here looks like you are testing a new thing (ref = ...), so it's best to split this test into a new:
```
 it("makes sure the '. ' is automatically added to publisher.", () =>{
    const ref =  docDefault.querySelector("#bib-TestRef2 + dd");
    // and so on... 
 })
```

File: tests/spec/core/biblio-spec.js [hunk 20]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
...
     expect(ref).toBeTruthy();
     expect(ref.textContent).toMatch(/Testing 123\.\s/);
     ref = null;
     // Make sure publisher is shown even when there is no author
-    ref = doc.querySelector("#bib-TestRef3 + dd");
+    ref = docDefault.querySelector("#bib-TestRef3 + dd");
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: Same here... it should be: 
```JS
it("makes sure publisher is shown even when there is no author", ()=>{
  const ref = docDefault.querySelector("#bib-TestRef3 + dd");
  // and so on...
})
```

File: tests/spec/core/biblio-spec.js [hunk 21]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
...
+    expect(ref.textContent.trim()).toEqual(finalString); 
+    //Make sure that the right things are hyperlinked
+    let anchors = ref.querySelectorAll("a");
+    expect(anchors.length).toEqual(2);
+    let [title, retrievedFrom] = [...anchors];
+    expect(title.href).toEqual("http://test.com/");
+    expect(title.firstElementChild.localName).toEqual("cite");
+    expect(title.firstElementChild.textContent).toEqual("Test ref title");
+    expect(retrievedFrom.href).toEqual("http://test.com/");
+    ref = finalString = null;
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: As above... these can be slit.

File: tests/spec/core/biblio-spec.js [hunk 22]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
...
+    expect(ref.textContent.trim()).toEqual(finalString); 
+    //Make sure that the right things are hyperlinked
+    let anchors = ref.querySelectorAll("a");
+    expect(anchors.length).toEqual(2);
+    let [title, retrievedFrom] = [...anchors];
+    expect(title.href).toEqual("http://test.com/");
+    expect(title.firstElementChild.localName).toEqual("cite");
+    expect(title.firstElementChild.textContent).toEqual("\"Test ref title.\"");
+    expect(retrievedFrom.href).toEqual("http://test.com/");
+    ref = finalString = null;
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: As above, split into 3 or more tests

File: tests/spec/core/biblio-spec.js [hunk 23]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
...
+    let ref = docAPA.querySelector("#bib-RefWithOnlyHref + dd");
+    expect(ref).toBeTruthy();
+    let finalString = "(2013-12-17). Retrieved from URL: http://test.com/";
+    expect(ref.textContent.trim()).toEqual(finalString); 
+    //Make sure that the right things are hyperlinked
+    let anchors = ref.querySelectorAll("a");
+    expect(anchors.length).toEqual(1);
+    [retrievedFrom] = [...anchors];
+    expect(retrievedFrom.href).toEqual("http://test.com/");
+    ref = finalString = null;
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: as above

File: tests/spec/core/biblio-spec.js [hunk 24]
```diff
@@ -92,26 +117,148 @@ describe("W3C — Bibliographic References", () => {
...
+    let ref = docAPA.querySelector("#bib-RefWithOnlyHrefAndAuthor + dd");
+    expect(ref).toBeTruthy();
+    finalString = "William Shakespeare. (2013-12-17). Retrieved from URL: http://test.com/";
+    expect(ref.textContent.trim()).toEqual(finalString); 
+    //Make sure that the right things are hyperlinked
+    anchors = ref.querySelectorAll("a");
+    expect(anchors.length).toEqual(1);
+    [retrievedFrom] = [...anchors];
+    expect(retrievedFrom.href).toEqual("http://test.com/");
+    ref = finalString = null;
```
Comment (marcoscaceres) [inline tests/spec/core/biblio-spec.js]: As above

**Comment by OrionStar25** [comment]
@marcoscaceres all the docs (docDefault, docAPA, docMLA) show undefined even when there is an async/await placed there. :scream:

**Comment by marcoscaceres** [comment]
These are starting to look good, @OrionStar25. I'll poke around a bit and see if I can get things unblocked.

**Comment by marcoscaceres** [comment]
@OrionStar25, this is totally not your fault, but I'm looking at APA and MLA citations styles, and now I'm worried that we can't adequately support them for a number of reasons - and they are both not suitable for technical standards. 

With [APA](http://www.bibme.org/citation-guide/apa/website/), the structure is:

> Author, A. (Year, Month Date of Publication). Article title. Retrieved from URL
 
1. We currently don't convert authors into the correct format, that is - Last-name, First Initial. There are also a bunch of rules around 3-7 Authors, and having to put a "&" in there.
1. The publisher is missing. 
1. The document status is missing.

The last two are critical, because they establish how reputable a citation is in the standards world. Similarly with MLA, the author format we have currently implemented is incorrect. In both cases, the citation styles are too limited to cite standards.

Furthermore, we don't adhere to the rules for inline citations for either MLA or APA. That is, we support: `As stated in [SomeReference]`, but MLA and APA require: `As stated in (Lastname, 2012)` - so we would end up with two different systems. 

Thus, I think we should not support additional bibliographical styles. I'm sorry I asked you to do this work without investigating first if the citation formats were actually suitable to use. I'd assumed they were, as they are widely used. But sadly, it turns out they are not very good :(

**Comment by OrionStar25** [comment]
Oh!! 🙈  Alright so we shouldn't go forward with this. That's cool, I
learnt a lot in this PR!

Well then can you please assign me another bug because I might choose
something way beyond my reach again? 😅

**Comment by marcoscaceres** [comment]
>Oh!! 🙈  Alright so we shouldn't go forward with this. That's cool, I learnt a lot in this PR!

Yeah, will close. 

> Well then can you please assign me another bug because I might choose something way beyond my reach again? 😅

I don't think this was beyond your reach. You were doing great - and I also didn't realize this until super late. This happens sometimes... more often than not! I often end up throwing out a lot of code because I realize too late things are not going to work. 

I'll assign you another one.

**Comment by OrionStar25** [comment]
Hey,

I just signed the petition "Students: Stop the rechristening of IIIT
Allahabad into IIIT Prayagraj" and wanted to see if you could help by
adding your name.

Our goal is to reach 5 signatures and we need more support. You can read
more and sign the petition here:

http://chng.it/Hw2scpLSpc

Thanks!
Niharika

**Pull request closed without merging**